scene/resources/material.cpp

```diff
@@ -700,252 +700,298 @@ void BaseMaterial3D::_update_shader() {
 	}
 
 	if (proximity_fade_enabled) {
-		code += "uniform sampler2D depth_texture : hint_depth_texture, repeat_disable, filter_nearest;";
+		code += "uniform sampler2D depth_texture : hint_depth_texture, repeat_disable, filter_nearest;\n";
 	}
 
 	if (features[FEATURE_NORMAL_MAPPING]) {
-		code += "uniform sampler2D texture_normal : hint_roughness_normal," + texfilter_str + ";\n";
-		code += "uniform float normal_scale : hint_range(-16,16);\n";
+		code += R"(
+uniform sampler2D texture_normal : hint_roughness_normal, )" +
+				texfilter_str + R"(;
+uniform float normal_scale : hint_range(-16.0, 16.0);
+)";
 	}
 	if (features[FEATURE_RIM]) {
-		code += "uniform float rim : hint_range(0,1);\n";
-		code += "uniform float rim_tint : hint_range(0,1);\n";
-		code += "uniform sampler2D texture_rim : hint_default_white," + texfilter_str + ";\n";
+		code += R"(
+uniform float rim : hint_range(0.0, 1.0, 0.01);
+uniform float rim_tint : hint_range(0.0, 1.0, 0.01);
+uniform sampler2D texture_rim : hint_default_white, )" +
+				texfilter_str + ";\n";
 	}
 	if (features[FEATURE_CLEARCOAT]) {
-		code += "uniform float clearcoat : hint_range(0,1);\n";
-		code += "uniform float clearcoat_roughness : hint_range(0,1);\n";
-		code += "uniform sampler2D texture_clearcoat : hint_default_white," + texfilter_str + ";\n";
+		code += R"(
+uniform float clearcoat : hint_range(0.0, 1.0, 0.01);
+uniform float clearcoat_roughness : hint_range(0.0, 1.0, 0.01);
+uniform sampler2D texture_clearcoat : hint_default_white, )" +
+				texfilter_str + ";\n";
 	}
 	if (features[FEATURE_ANISOTROPY]) {
-		code += "uniform float anisotropy_ratio : hint_range(0,256);\n";
-		code += "uniform sampler2D texture_flowmap : hint_anisotropy," + texfilter_str + ";\n";
+		code += R"(
+uniform float anisotropy_ratio : hint_range(0.0, 1.0, 0.01);
+uniform sampler2D texture_flowmap : hint_anisotropy, )" +
+				texfilter_str + ";\n";
 	}
 	if (features[FEATURE_AMBIENT_OCCLUSION]) {
-		code += "uniform sampler2D texture_ambient_occlusion : hint_default_white, " + texfilter_str + ";\n";
-		code += "uniform vec4 ao_texture_channel;\n";
-		code += "uniform float ao_light_affect;\n";
+		code += R"(
+uniform sampler2D texture_ambient_occlusion : hint_default_white, )" +
+				texfilter_str + R"(;
+uniform vec4 ao_texture_channel;
+uniform float ao_light_affect : hint_range(0.0, 1.0, 0.01);
+)";
 	}
 
 	if (features[FEATURE_DETAIL]) {
-		code += "uniform sampler2D texture_detail_albedo : source_color," + texfilter_str + ";\n";
-		code += "uniform sampler2D texture_detail_normal : hint_normal," + texfilter_str + ";\n";
-		code += "uniform sampler2D texture_detail_mask : hint_default_white," + texfilter_str + ";\n";
+		code += R"(
+uniform sampler2D texture_detail_albedo : source_color, )" +
+				texfilter_str + R"(;
+uniform sampler2D texture_detail_normal : hint_normal, )" +
+				texfilter_str + R"(;
+uniform sampler2D texture_detail_mask : hint_default_white, )" +
+				texfilter_str + ";\n";
```

Comment:
There are places like these in the PR where I do question if raw strings are the way to go. Sure, it's consistent acriss the board, but it's a bit more "incomprehensible".